compiler/rustc_target/src/abi/mod.rs

```diff
@@ -677,32 +677,59 @@ impl Primitive {
     }
 }
 
+/// Inclusive wrap-around range of valid values, that is, if
+/// start > end, it represents `start..=MAX`,
+/// followed by `0..=end`.
+///
+/// That is, for an i8 primitive, a range of `254..=2` means following
+/// sequence:
+///
+///    254 (-2), 255 (-1), 0, 1, 2
+///
+/// This is intended specifically to mirror LLVM’s `!range` metadata,
+/// semantics.
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+#[derive(HashStable_Generic)]
+pub struct WrappingRange {
+    pub start: u128,
+    pub end: u128,
+}
+
+impl WrappingRange {
+    /// Returns `true` if `v` is contained in the range.
+    #[inline]
+    pub fn contains(&self, v: u128) -> bool {
+        if self.start <= self.end {
+            self.start <= v && v <= self.end
+        } else {
+            self.start <= v || v <= self.end
+        }
+    }
+
+    /// Returns `true` if zero is contained in the range.
+    /// Equal to `range.contains(0)` but should be faster.
+    #[inline]
+    pub fn contains_zero(&self) -> bool {
+        !(self.start <= self.end && self.start != 0)
```

Comment:
IMO `self.start > self.end || self.start == 0` would be a lot more readable.

Reply:
Yeah, wrote it originally as is_non_null, changing it

Reply:
Is `start < end` right?, shouldn't it be `start <= end`

Reply:
`start > end` means it wraps around which means it does contain 0.

Might be good to have a comment explaining that. :)

Reply:
Just checking if you think that the code before my PR was incorrect?

Reply:
The code I proposed is equivalent to your `contains_zero`.

I assume you meant to add a negation? So that would be `!(self.start < self.end && self.start != 0)` for the old code which is equivalent to `self.start >= self.end || self.start == 0`. Yeah that seems incorrect to me; the range `5..=5` does not contain 0.

Reply:
great thanks!